larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewDeltaRayMatchingAlgorithm.cc

```diff
@@ -0,0 +1,466 @@
+
+    ProtoParticleVector protoParticleVector({protoParticle});
+
+    return (this->CreatePfos(protoParticleVector));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TwoViewDeltaRayMatchingAlgorithm::FormThirdViewCluster(const MatrixType::Element &element, ProtoParticle &protoParticle)
+{
+    const PfoList &commonMuonPfoList(element.GetOverlapResult().GetCommonMuonPfoList());
+    const Cluster *const pBestMatchedCluster(element.GetOverlapResult().GetBestMatchedCluster());
+    const HitType &thirdViewHitType(LArClusterHelper::GetClusterHitType(pBestMatchedCluster));
+    const ParticleFlowObject *pMatchedMuonPfo(nullptr);
+    
+    for (const ParticleFlowObject *const pMuonPfo : commonMuonPfoList)
+    {
+        ClusterList muonClusterList;
+        LArPfoHelper::GetClusters(pMuonPfo, thirdViewHitType, muonClusterList);
+
+        if (std::find(muonClusterList.begin(), muonClusterList.end(), pBestMatchedCluster) != muonClusterList.end())
+            pMatchedMuonPfo = pMuonPfo;
+    }
+
+    const Cluster *pThirdViewCluster(pMatchedMuonPfo ? nullptr : pBestMatchedCluster);
+    
+    if (pMatchedMuonPfo)
+    {
+        CaloHitList deltaRayHitList;
+        
+        if (this->CollectHitsFromMuon(element.GetCluster1(), element.GetCluster2(), nullptr, pMatchedMuonPfo, 
+            m_minDistanceFromMuon, m_maxDistanceToCollected, deltaRayHitList) == STATUS_CODE_SUCCESS)
+        {
+            this->SplitMuonCluster(this->GetThirdViewClusterListName(), pBestMatchedCluster, deltaRayHitList, pThirdViewCluster);
+            this->UpdateForThirdViewClusterModification(pBestMatchedCluster, true);
+        }
+        else
+        {
+            pThirdViewCluster = element.GetOverlapResult().GetBestMatchedAvailableCluster();
+        }
+    }
+    
+    if (!pThirdViewCluster)
+        return;
+
+    this->UpdateForThirdViewClusterModification(pThirdViewCluster, false);
+    this->MergeThirdView(element, pThirdViewCluster);
+
+    protoParticle.m_clusterList.push_back(pThirdViewCluster);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TwoViewDeltaRayMatchingAlgorithm::MergeThirdView(const MatrixType::Element &element, const Cluster *const pSeedCluster)
+{
+    CaloHitList caloHitList1, caloHitList2;
+    element.GetCluster1()->GetOrderedCaloHitList().FillCaloHitList(caloHitList1);
+    element.GetCluster2()->GetOrderedCaloHitList().FillCaloHitList(caloHitList2);
+
+    // ATTN: Need copy as original will change throughout function 
+    ClusterList matchedClusters(element.GetOverlapResult().GetMatchedClusterList());
+        
+    ClusterSet checkedClusters;
+    checkedClusters.insert(pSeedCluster);
+        
+    while (checkedClusters.size() != matchedClusters.size())
```

Comment:
Unless something has gone wrong, the answer is no. For this function to be called the pBestCluster has to not be a nullptr which means that the matchedClusters list has to have at least one element. The function UpdateForThirdViewClusterModification does all the cleanup of the matchedClusters list and the pBestMatchedCluster whenever a pfo is created. So, you have to let the underlying matrix update between pfo creations to prevent reading 'outdated' matched clusters lists? Which is why (when you come to it) in the TwoViewUnambiguousDR alg i have to create pfos one at a time.

However, you made me realise something.. I really hate this loop, essentially i am trying to get every element in matchedClusters checked but in a specific order dictated by their hit number. I try to stop the pSeedCluster from being evaluated since i dont want to merge something with itself so I add it into the checkedCluster set beforehand. However, i realised that the pSeedCluster isn't always in that list since, in the case of muons, I can actually create a new cluster which is then the pSeedCluster => I changed it to:

`    if (checkedClusters.count(pSeedCluster))
        checkedClusters.insert(pSeedCluster);`